File: problems_051to075/problem_051/solution.py
import os
import sys

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch
from toolbox import is_prime_naive


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = None

        prime_count_per_pattern = {}
        num = 9

        while answer is None:
            num += 2

            if not is_prime_naive(num):
                continue

            temp = str(num)
            unique_digits = set(temp[:-1])

            for digit in unique_digits:
                generating_pattern = temp[:-1].replace(digit, '*') + temp[-1:]

                if generating_pattern in prime_count_per_pattern:
                    continue

                prime_count = 0

                for substitute in '0123456789':
                    decimal = generating_pattern.replace('*', substitute)

                    if decimal[0] == '0':
                        continue

                    if is_prime_naive(int(decimal)):
                        prime_count += 1

                if prime_count == 8:
                    answer = num
                    break

                prime_count_per_pattern[generating_pattern] = prime_count

        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_051to075/problem_051/solution.ts
import { isPrime } from 'mathjs';

import { SolutionFunction, Dictionary } from '../../lib/types';

const solution: SolutionFunction = () => {
  let answer: ReturnType<SolutionFunction> = null;

  const primeCountPerPattern: Dictionary<number> = {};
  let num = 9;

  while (answer === null) {
    num += 2;

    if (isPrime(num)) {
      const temp = String(num);
      const uniqueDigits = [...new Set(temp.slice(0, -1))];

      for (let index = 0; index < uniqueDigits.length; index += 1) {
        const digit = uniqueDigits[index];
        const generatingPattern = temp.slice(0, -1).replace(new RegExp(digit, 'g'), '*') + temp.slice(-1);

        if (!(generatingPattern in primeCountPerPattern)) {
          let primeCount = 0;

          '0123456789'.split('').forEach((substitute) => {
            const decimal = generatingPattern.replace(/\*/g, substitute);

            if (decimal[0] !== '0') {
              if (isPrime(Number(decimal))) {
                primeCount += 1;
              }
            }
          });

          if (primeCount === 8) {
            answer = num;
            break;
          }

          primeCountPerPattern[generatingPattern] = primeCount;
        }
      }
    }
  }

  return answer;
};

export default solution;
